### PR TITLE
[refactor][utils] Create utils for component manager

### PR DESCRIPTION
**Pull request by SHtress**
Move frequently used functions to componentUtils::SearchUtils
- GetComponentAddress
- GetComponentDependencies
- GetComponentInstallationMethod

### CODE REVIEW

File: src/manager/component_utils/sc_component_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * @brief Get component dependencies vector
+ * @param context current sc-memory context
+ * @param componentAddr sc-addr of the component to install
+ * @return ScAddrVector consists of component dependencies sc-addrs,
+ * return empty vector if component has no dependencies
+ */
+ScAddrVector ComponentUtils::GetComponentDependencies(ScMemoryContext * context, ScAddr const & componentAddr)
+{
+  ScAddrVector componentDependencies;
+  ScIterator5Ptr const componentDependenciesIterator = context->Iterator5(
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.cpp:48]:
```suggestion
  ScIterator5Ptr const & componentDependenciesIterator = context->Iterator5(
```

File: src/manager/component_utils/sc_component_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      keynodes::ScComponentManagerKeynodes::nrel_component_dependencies);
+
+  while (componentDependenciesIterator->Next())
+  {
+    ScAddr componentDependenciesSet = componentDependenciesIterator->Get(2);
+    ScAddrVector componentCurrentDependencies =
+        utils::IteratorUtils::getAllWithType(context, componentDependenciesSet, ScType::NodeConst);
+    componentDependencies.insert(
+        componentDependencies.cend(), componentCurrentDependencies.cbegin(), componentCurrentDependencies.cend());
+  }
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.cpp:62]: Move declaration from cycle
```suggestion
  ScAddr componentDependenciesSet;
  ScAddrVector componentCurrentDependencies;
  while (componentDependenciesIterator->Next())
  {
    componentDependenciesSet = componentDependenciesIterator->Get(2);
    componentCurrentDependencies =
        utils::IteratorUtils::getAllWithType(context, componentDependenciesSet, ScType::NodeConst);
    componentDependencies.insert(
        componentDependencies.cend(), componentCurrentDependencies.cbegin(), componentCurrentDependencies.cend());
  }
```

File: src/manager/component_utils/sc_component_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  ScAddr componentInstallationMethod;
+  ScIterator5Ptr const componentDependenciesIterator = context->Iterator5(
+      componentAddr,
+      ScType::EdgeDCommonConst,
+      ScType::NodeConst,
+      ScType::EdgeAccessConstPosPerm,
+      keynodes::ScComponentManagerKeynodes::nrel_installation_method);
+
+  if (componentDependenciesIterator->Next())
+    componentInstallationMethod = componentDependenciesIterator->Get(2);
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.cpp:84]:
```suggestion
  if (componentDependenciesIterator->Next())
  {
    componentInstallationMethod = componentDependenciesIterator->Get(2);
  }
```

File: src/manager/component_utils/sc_component_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  return componentInstallationMethod;
+}
+}  // namespace utils
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.cpp:88]:
```suggestion
}

}  // namespace utils

```

File: src/manager/component_utils/sc_component_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <sc-memory/sc_memory.hpp>
+
+namespace utils
+{
+class ComponentUtils
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.hpp:7]:
```suggestion
{

class ComponentUtils
```

File: src/manager/component_utils/sc_component_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+{
+public:
+  static std::string GetComponentAddress(ScMemoryContext * context, ScAddr const & componentAddr);
+
+  static ScAddrVector GetComponentDependencies(ScMemoryContext * context, ScAddr const & componentAddr);
+
+  static ScAddr GetComponentInstallationMethod(ScMemoryContext * context, ScAddr const & componentAddr);
+};
+}  // namespace utils
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.hpp:16]:
```suggestion
};

}  // namespace utils

```

File: src/manager/component_utils/sc_component_utils.hpp (hunk 5 above)
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.hpp:7]: I'd like to name it like "SearchUtils". Or smth like that to identify that there are "GetSmth" methods

File: src/manager/component_utils/sc_component_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <sc-memory/sc_memory.hpp>
+
+namespace utils
```
Comment (MksmOrlov) [inline src/manager/component_utils/sc_component_utils.hpp:5]: I'd like to name it like "componentUtils"

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
I suggest to name directory as "utils", not "component_utils"

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 8]
```diff
@@ -8,8 +8,7 @@
 
 #include "sc-memory/utils/sc_exec.hpp"
 #include "../sc-builder/src/scs_loader.hpp"
-#include "sc-agents-common/utils/CommonUtils.hpp"
-#include "sc-agents-common/utils/IteratorUtils.hpp"
+#include "src/manager/component_utils/sc_component_utils.hpp"
```
[the remark this answers is not in the document]
Comment (SHtress) [reply src/manager/commands/command_install/sc_component_manager_command_install.cpp:11]: For now we need loader for installation after download, but in further pr after installation method update i remove scs_loader

File: src/manager/component_utils/sc_component_utils.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+namespace utils
+{
+
+/**
+ * @brief Get content of component address link content
+ * @param context current sc-memory context
+ * @param componentAddr sc-addr of the component to install
+ * @return string of component address
+ */
+std::string ComponentUtils::GetComponentAddress(ScMemoryContext * context, ScAddr const & componentAddr)
```
[the remark this answers is not in the document]
Comment (SHtress) [reply src/manager/component_utils/sc_component_utils.cpp:18]: Update to just take content in install method

File: src/manager/component_utils/sc_component_utils.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      componentAddr,
+      ScType::EdgeDCommonConst,
+      ScType::LinkConst,
+      ScType::EdgeAccessConstPosPerm,
+      keynodes::ScComponentManagerKeynodes::nrel_component_address);
+
+  if (componentAddressIterator->Next())
+  {
+    componentAddressAddr = componentAddressIterator->Get(2);
+    componentAddress = utils::CommonUtils::getLinkContent(context, componentAddressAddr);
```
[the remark this answers is not in the document]
Comment (SHtress) [reply src/manager/component_utils/sc_component_utils.cpp:32]: Context method returns ScStream and method from utils just validates it and converts to string

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 11]
```diff
@@ -226,29 +145,30 @@ ExecutionResult ScComponentManagerCommandInstall::InstallDependencies(
...
     {
     };
-    size_t componentDirNameIndex = componentAddress.rfind('/');
-    std::string componentDirName = m_specificationsPath + componentAddress.substr(componentDirNameIndex);
+    size_t componentDirNameIndex = componentAddressContent.rfind('/');
+    std::string componentDirName = m_specificationsPath + componentAddressContent.substr(componentDirNameIndex);
     while (stat(componentDirName.c_str(), &sb) == 0)
     {
-      componentDirName += componentAddress.substr(componentDirNameIndex);
+      componentDirName += componentAddressContent.substr(componentDirNameIndex);
```
Comment (MksmOrlov) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:160]: use stringstream. Do not concatenate strings

File: src/manager/utils/sc_component_utils.cpp [hunk 12]
```diff
@@ -0,0 +1,93 @@
+
+
```
Comment (MksmOrlov) [inline src/manager/utils/sc_component_utils.cpp:2]: empty lines. add licence header

File: src/manager/utils/sc_component_utils.hpp [hunk 13]
```diff
@@ -0,0 +1,18 @@
+#pragma once
```
Comment (MksmOrlov) [inline src/manager/utils/sc_component_utils.hpp:1]: add licence header

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp [hunk 14]
```diff
@@ -171,15 +87,17 @@ bool ScComponentManagerCommandInstall::ValidateComponent(ScMemoryContext * conte
   }
 
   // Find and check component address
-  std::string const componentAddress = GetComponentAddress(context, componentAddr);
-  if (componentAddress.empty())
+  ScAddr const & componentAddressAddr = componentUtils::SearchUtils::GetComponentAddress(context, componentAddr);
+  std::string const & componentAddressContent = utils::CommonUtils::getLinkContent(context, componentAddressAddr);
```
Comment (MksmOrlov) [inline src/manager/commands/command_install/sc_component_manager_command_install.cpp:91]: use ScMemoryContext method to get link content

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: src/manager/component_utils/sc_component_utils.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (MksmOrlov) [reply src/manager/component_utils/sc_component_utils.cpp]: No, use another method

File: src/manager/commands/command_install/sc_component_manager_command_install.cpp (hunk 11 above)
Comment (SHtress) [reply src/manager/commands/command_install/sc_component_manager_command_install.cpp:160]: Should we fix it now cause next pr will update whole download and install methods?

**Review by MksmOrlov — APPROVED** [review]

**Review by smousy — APPROVED** [review]

**Pull request merged**